torch/autograd/gradcheck.py

```diff
@@ -97,14 +95,19 @@ def get_stride(size):
                     r = (outb - outa) / (2 * eps)
                     d_tensor[d_idx] = r.detach().reshape(-1)
         else:
-            x_tensor = x_tensor.data
             for d_idx, x_idx in enumerate(product(*[range(m) for m in x_tensor.size()])):
                 orig = x_tensor[x_idx].item()
-                x_tensor[x_idx] = orig - eps
+                # We don't want in-place changes to `x_tensor` to update its version
+                # counter, because call sites of `get_numerical_jacobian(...)` expects
+                # `input`'s version counter(s) to not be changed by the function call.
+                with torch.no_grad(update_version=False):
+                    x_tensor[x_idx] = orig - eps
                 outa = fn(input).clone()
-                x_tensor[x_idx] = orig + eps
+                with torch.no_grad(update_version=False):
+                    x_tensor[x_idx] = orig + eps
                 outb = fn(input).clone()
-                x_tensor[x_idx] = orig
+                with torch.no_grad(update_version=False):
+                    x_tensor[x_idx] = orig
```

Comment:
I would like to propose the `with torch.no_grad(update_version=False)` API, because this function (`get_numerical_jacobian(...)`, which is a crucial function used by `gradcheck()` and `gradgradcheck()`) wants the original version counter values of `x_tensor` to always be preserved, so that making small finite changes to `x_tensor` (and restoring its original value later) doesn't invalidate it in the autograd graph.

Reply:
OK, this looks sound to me.

I might explain what's going on this code a little differently. Basically, the invariant required by version is that we must update the version if the contents of the tensor change. We are preserving this invariant because we reset the tensor to its original value after running the finite differences, but PyTorch doesn't know that so we help it out.

Your comment says that it's important not to update the version counter, but it doesn't explain why *not* updating the version counter is sound. That's an important thing to write down.